File: pyspider/feature_gates.py
from __future__ import annotations

import os


_DEFAULTS = {
    "ai": True,
    "browser": True,
    "distributed": True,
    "media": True,
    "workflow": True,
    "crawlee": True,
}

_PROFILES = {
    "lite": ["browser", "workflow"],
    "ai": ["ai", "browser", "workflow"],
    "distributed": ["browser", "distributed", "workflow", "crawlee"],
    "full": ["ai", "browser", "distributed", "media", "workflow", "crawlee"],
}


def _env_name(name: str) -> str:
    return f"PYSPIDER_FEATURE_{name.upper().replace('-', '_').replace('.', '_')}"


def is_enabled(name: str) -> bool:
    normalized = name.strip().lower()
    raw = os.getenv(_env_name(normalized), "").strip().lower()
    if not raw:
        return _DEFAULTS.get(normalized, False)
    return raw in {"1", "true", "yes", "on"}


def catalog() -> dict[str, object]:
    return {
        "default_profile": "full",
        "profiles": _PROFILES,
        "env_prefix": "PYSPIDER_FEATURE_",
        "features": {name: is_enabled(name) for name in sorted(_DEFAULTS)},
    }

File: rustspider/src/feature_gates.rs
use serde_json::{json, Value};
use std::collections::BTreeMap;
use std::env;

fn default_state() -> BTreeMap<&'static str, bool> {
    BTreeMap::from([
        ("ai", true),
        ("browser", true),
        ("distributed", true),
        ("media", true),
        ("workflow", true),
        ("crawlee", true),
    ])
}

pub fn enabled(name: &str) -> bool {
    let normalized = name.trim().to_ascii_lowercase();
    let env_name = format!(
        "RUSTSPIDER_FEATURE_{}",
        normalized.replace(['-', '.'], "_").to_ascii_uppercase()
    );
    match env::var(&env_name) {
        Ok(value) => matches!(
            value.trim().to_ascii_lowercase().as_str(),
            "1" | "true" | "yes" | "on"
        ),
        Err(_) => default_state()
            .get(normalized.as_str())
            .copied()
            .unwrap_or(false),
    }
}

pub fn catalog() -> Value {
    let defaults = default_state();
    json!({
        "default_profile": "full",
        "profiles": {
            "lite": ["browser", "workflow"],
            "ai": ["ai", "browser", "workflow"],
            "distributed": ["browser", "distributed", "workflow", "crawlee"],
            "full": ["ai", "browser", "distributed", "media", "workflow", "crawlee"]
        },
        "env_prefix": "RUSTSPIDER_FEATURE_",
        "features": defaults
            .keys()
            .map(|name| ((*name).to_string(), enabled(name)))
            .collect::<BTreeMap<String, bool>>()
    })
}
